src/handle/command.ml

```diff
@@ -407,31 +407,79 @@ let handle : (Path.t -> Sign.t) -> sig_state -> p_command ->
             (* Just ignore the command with a warning. *)
             wrn pe.pos "Proof aborted."; ss
         | _ ->
-            (* Check that no metavariable remains. *)
-            if LibTerm.Meta.has true a then
-              (fatal_msg "The type of %a has unsolved metavariables.\n"
-                 pp_uid id;
-               fatal pe.pos "We have %a : %a." pp_uid id pp_term a);
-            (match t with
-             | Some(t) when LibTerm.Meta.has true t ->
-                 fatal_msg
-                   "The definition of %a has unsolved metavariables.\n"
-                   pp_uid id;
-                 fatal pe.pos "We have %a : %a ≔ %a."
-                   pp_uid id pp_term a pp_term t
-             | _ -> ());
             match pe.elt with
             | P_proof_abort -> assert false (* Handled above *)
             | P_proof_admit ->
                 (* If the proof is finished, display a warning. *)
                 if finished ps then
                   wrn pe.pos "The proof is finished. Use 'end' instead.";
+                (* Get a list of metas, ordered by dependency. *)
+                let metas =
+                  let type_m = LibTerm.Meta.get true a in
+                  let metas =
+                    match t with
+                    | Some t -> MetaSet.union type_m (LibTerm.Meta.get true t)
+                    | None -> type_m
+                  in
+                  let dep m_a m_b =
+                    (* If [m_a] depends on [m_b], then it is greater than
+                       [m_b], else they are equal. *)
+                    if LibTerm.Meta.depends m_a m_b then 1 else 0
+                  in
+                  List.stable_sort dep (MetaSet.to_seq metas |> List.of_seq)
+                in
+                (* [add_axiom ss m] adds an axiom for meta [m] in signature
+                   state [ss]. *)
+                let add_axiom ss m =
+                  let name =
+                    "ax_" ^ Option.get (string_of_int m.meta_key) m.meta_name
```

Comment:
Two axioms could get the same name as the meta key counter is reset at each command. You should use a specific counter for each file instead.
But, even with this, such a name could be in conflict with a user name introduced before or later. I see no solution at the moment. For the moment, perhaps use a prefix less likely to be used like "__ax_".